README
=begin
= RuBBS 1.0.7

== RuBBSとは
RuBBSは、Rubyで書かれた掲示板CGIスクリプトです。以下のような特徴があります。

* 複数の掲示板を管理できます。
* 設定の変更は、すべて管理用のCGIを通じて行なえます。
* 最新記事の参照は静的に生成されたHTMLを使うため、非常に負荷が低いです。
* 投稿の内容を、任意の複数のメールアドレスに配信できます。
* Cookieを使っていくつかの情報を保存します。
* eRubyを使った柔軟なカスタマイズが可能です。
* 記事中のURLは自動的にリンクに変換されます。一方、HTMLタグは使用できないので安全です。

== インストール
すでに0.3.2までのRuBBSを使っていた場合は、このあとの『移行について』も読んでください。

まず、RuBBSをインストールするディレクトリを作成します。
CGIスクリプトを入れるディレクトリと、静的HTMLファイルを格納するディレクトリの2つが必要です。仮に以下のように作成したとします。

* CGIスクリプト用: /var/www/cgi-bin/rubbs
* HTML用: /var/www/html/rubbs

それぞれに、WWWサーバの権限で書き込みができるようにしておく必要があります
(ownerを変更する、chmod o+wするなど)。

続いてCGIスクリプトのインストールです。

配布されたファイルはすべて、CGIスクリプト用ディレクトリにコピーします。
以下のようになります。

  /var/www/cgi-bin/rubbs --- rubbs.rb
      +- rubbs.rb
      +- admin
      +- admin-init
      +- lock.rb
      +- erb/
      |   +- erb*.rb
      +- conf/
      |   +- *.conf
      |   +- *.rhtml
      |   +- *.rtxt
		+- misc/
		|   +- squeeze.*
		|   +- postrubbs.*
      +- README.*
      +- CHANGES
      +- COPYING

環境によってはこのままでも動きますが、
以下のような制限がある場合には少しスクリプトを書き換える必要があります。

(1) CGIのサフィックス(拡張子)が特定のもの(.cgiなど)でなければならない
    adminとadmin-initのサフィックスを適切なものに変更してください。
(2) Rubyのインストール先が特殊
    WWWサーバの検索パスに含まれていない場合などには、adminとadmin-initの最初の行を以下のように書き換える必要があります。

        #!/usr/bin/env ruby
                ↓
        #!/foo/bar/bin/ruby
        (/foo/bar/にRubyがインストールされている場合)

== 実行方法
Administration Tool(adminスクリプト)に、CGIとしてブラウザからアクセスします。
インストールがうまくいっていないとエラーが出るかも知れませんが、
adminからのメッセージの場合はそれに従って設定を変更すると、
管理用のパスワードや各種初期設定をする画面になるはずです。
ここまでは英語ですが、その後は日本語の画面になるので、我慢して下さい(笑)。
日本語にならない場合、使っているブラウザの設定が英語優先になっている可能性があります。
Accept-Languageで「ja」を優先するように設定してください。
もちろん、英語のままでも日本語対応の掲示板を作成できます。

初期設定が済んだら、adminは管理者用のパスワードの入力を促す画面になります。
今後の管理はこの画面からスタートすることになります。

あとは画面の指示に従って操作してください。

== 旧バージョンからの移行について
0.9.1以降のRuBBSは、それ以前に使われていたCSSと互換性が一部失われています。添付のconf/rubbs.cssを参考にして書き直してください。

0.5.0以降のRuBBSは、0.3.2以前のものとは設定ファイルやデータベースファイルの互換性がありません。
このため、以下の作業が必要です。

* CGIスクリプトのインストール先は別のディレクトリにした方が良いでしょう。
ただしHTMLファイルの格納先は同じの方が良いでしょう。
* 旧掲示板の内容は移行することができます。Administration Tool(adminスクリプト)で移行してください。
このメニューは、ブラウザのAccept-Languageが「ja」の時のみ現れます(旧RuBBSを英語圏で使っている人はいないため)。もしメニューが出なかった場合は、ブラウザの設定を見直してください。
* 設定は引き継ぐことができません。新たに掲示板を作り直してください。
* 旧版で変更可能だった色の設定はなくなりました。
見栄えはすべてCSSに出してありますので、自由に変更してください。
標準のCSSはconf/rubbs.cssにあります。このファイルは最初のadmin実行後、HTMLパスにコピーされます。

== ライセンス
GPL 2.0に準拠します。著作権は「ただただし」(sho@spc.gr.jp)が保有しています。
サポートはhttp://www.spc.gr.jp/bbs/RuBBS.htmlで行なっています。

ただし、付属のerb(ディレクトリerb配下のすべてのファイル)は、咳さんの作品です。
<http://www2a.biglobe.ne.jp/~seki/ruby/>を参照してください。
=end



ChangeLog
2003.09.26 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.7.
	* support ruby 1.8.0.

2002.07.26 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.6
	* conf/default.view.rhtml, conf/default.form.rhtml: escape HTML for mail and name.
	* rubbs.rb, conf/default.conf.rhtml: no re-generate HTML when call script without effect parameter.

2002.04.23 TADA Tadashi <sho@spc.gr.jp>
	* misc/squeeze.rb: 1.0.4

2002.03.19 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.5
	* admin: remove checking validation of @url.

2002.02.18 TADA Tadashi <sho@spc.gr.jp>
	* add title attr into links of reply_to.

2002.01.25 TADA Tadashi <sho@spc.gr.jp>
	* initial save @suffix in admin-init.

2002.01.16 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.4
	* remove CHANGES.
	* fix DOCTYPE.

2001.12.28 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.3
	* misc/squeeze.rb: 1.0.3. fix bug.

2001.12.25 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.2
	* lock.rb: Process::pid -> $$.
	* rename CHANGES -> ChangeLog.

2001.12.18 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.1
	* rubbs.rb, lock.rb: support Win32.

2001.11.25 TADA Tadashi <sho@spc.gr.jp>
	* version 1.0.0
	* default.view.rhtml: added <div class="article">.
	* default.conf.rhtml: rebuild html button.
	* *.rhtml: disable CSS on Netscape 4.x.

2001.11.23 TADA Tadashi <sho@spc.gr.jp>
	* version 0.9.4
	* misc/postrubbs.rb: trimming Subject treatment.

2001.11.21 TADA Tadashi <sho@spc.gr.jp>
	* version 0.9.3
	* misc/postrubbs.rb: trimming Subject and From treatment.
	* misc/postrubbs.rb: don't show e-mail address in default.

2001.11.19 TADA Tadashi <sho@spc.gr.jp>
	* version 0.9.2
	* rubbs.rb, conf/mail.rtxt: mail force to ISO-2022-JP when Japanese.
	* rubbs.rb, misc/postrubbs.rb: trimming Subject treatment.

2001.11.17 TADA Tadashi <sho@spc.gr.jp>
	* version 0.9.1
	* valid XHTML 1.0
	* valid CSS 1.0
	* misc/squeeze.rb: 1.0.2
	* misc/postrubbs.rb: 1.0.0

2001.09.13 TADA Tadashi <sho@spc.gr.jp>
	* version 0.9.0
	* fix bug about remove and restore.
	* In-Reply-To in mail.rtxt.
	* supress time data from message-id in mail.
	* added link to original article in subject line.
	* changing anchor of an artcle to "top=xxx".
	* link of next/prev articlas.
	* conf/default.one.rhtml: updating subject line.

2001.07.12 TADA Tadashi <sho@spc.gr.jp>
	* version 0.7.0
	* security fix: remove images in all rhtml files.
	* security fix: change default css path to local.
	* conf/mail.rtxt: MIME-Version on header.

2001.06.06 TADA Tadashi <sho@spc.gr.jp>
	* version 0.5.1
	* admin: change a wrong English message.
	* rubbs.rb: anchor of each articles.
	* rubbs.rb: quoted html bug fixed.
	* rubbs.rb: Cookie expire after 1 month.
	* rubbs.rb: some changes for Ruby 1.6.4.

2001.03.09 TADA Tadashi <sho@spc.gr.jp>
	* version 0.5.0
	* fully rewrite by eRuby(ERb).
	* support administration tool, and remove makebbs command.
	* change DB format DBM to GDBM. But if no gdbm, it uses DBM.
	* support CSS, and remove color settings.
	* support multi-charset, ISO-8859-1/ISO-2022-JP/EUC-JP/Shift_JIS.
	* suppress showing To field via E-mail.

2000.09.23 TADA Tadashi <sho@spc.gr.jp>
	* version 0.3.2
	* for Ruby 1.4.x or later (using cgi and net/smtp).
	* convert all input to Shift_JIS.
	* suppress MIME convert.

2000.05.06 TADA Tadashi <sho@spc.gr.jp>
	* version 0.3.1
	* WRAP="SOFT" in TEXTAREA disabled.
	* "Pragma: no-cache" deleted on view.
	* cookie specification changed. No TAB char in cookie.
	* some bugs fixed.

2000.05.04 by l <l@kondara.org>, TADA Tadashi <sho@spc.gr.jp>
	* version 0.3.0
	* quote color.
	* cookie support.
	* separated form option.

2000.02.27 TADA Tadashi <sho@spc.gr.jp>
	* version 0.2.4
	* fix wait bug of lock.
	* escape HTML tags in <TEXTAREA>.
	* force unlock after 5 min.

2000.02.16 TADA Tadashi <sho@spc.gr.jp>
	* version 0.2.3
	* fix problem space after URL.
	* reply button

1999.09.23 TADA Tadashi <sho@spc.gr.jp>
	* version 0.2.2
	* RuBBS::lock method.
	* Bug fix RuBBS::load.
	* Replaced Ruby Banner.

1999.09.09 TADA Tadashi <sho@spc.gr.jp>
	* version 0.2.1
	* replacing white space with '&nbsp;'.
	* undo delete.

1999.09.03 TADA Tadashi <sho@spc.gr.jp>
	* version 0.2.0
	* for Ruby 1.2/1.3/1.4.
	* killing HTML tag for subject and name.

lock.rb
#
# Lock module
#
# Copyright (C) 2001, All right reserved by TADA Tadashi <sho@spc.gr.jp>
# You can redistribute it and/or modify it under GPL2.
#
module Lock
	class LockError < StandardError; end

	def Lock::lock( basefile = 'lock/lock', try = 10, timeout = 600 )
		locking = false
		lockfile = "#{basefile}.#$$.#{Time.now.to_i}"
		try.times do
			begin
				File.rename( basefile, lockfile )
				locking = true
				break
			rescue #rename failed
				begin
					Dir.glob( "#{basefile}.*" ).each do |oldlock|
						if /^#{basefile}\.\d+\.(\d+)/ =~ oldlock and
								Time.now.to_i - $1.to_i > timeout then
							File.rename( oldlock, lockfile )
							locking = true
							break
						end
					end
				rescue
				end
			end
			sleep( 1 )
		end
		raise Lock::LockError, "LockError: #{lockfile} #{$!}" unless locking
		if iterator? then
			begin
				yield
			ensure
				Lock::unlock( lockfile, basefile ) if locking
			end
		else
			return lockfile
		end
	end

	def Lock::unlock( lockfile, basefile = 'lock/lock' )
		File.rename( lockfile, basefile )
	end
end


erb/erbl.rb
# Tiny Tiny eRuby --- ERbLight
# 	Copyright (c) 1999-2000 Masatoshi SEKI 
#       You can redistribute it and/or modify it under the same term as Ruby.
# 	$Id: erbl.rb,v 1.1.1.1 2001/06/08 08:02:25 spc Exp $
# 	$Author: spc $

require 'erb/compile'

class ERbLight
  Revision = '$Date: 2001/06/08 08:02:25 $'

  def self.version
    "erbl.rb [1.3.3 #{ERbLight::Revision.split[1]}]"
  end

  def initialize(str, safe_level=nil, trim_mode=nil, eoutvar='_erbout')
    @safe_level = safe_level
    compiler = ERbCompiler.new
    compiler.trim_mode = trim_mode
    set_eoutvar(compiler, eoutvar)
    @src = compiler.compile(str)
  end
  attr :src

  def set_eoutvar(compiler, eoutvar = '_erbout')
    compiler.put_cmd = "#{eoutvar}.concat"

    cmd = []
    cmd.push 'begin'
    cmd.push("$SAFE = #{@safe_level}") if @safe_level
    cmd.push "#{eoutvar} = ''"
    
    compiler.pre_cmd = cmd

    cmd = []
    cmd.push('ensure')
    cmd.push(eoutvar)
    cmd.push('end')

    compiler.post_cmd = cmd
  end

  def run(b=TOPLEVEL_BINDING)
    print self.result(b)
  end

  def result(b=TOPLEVEL_BINDING)
    if @safe_level
      th = Thread.start { 
	eval(@src, b)
      }
      return th.value
    else
      return eval(@src, b)
    end
  end
end

if __FILE__ == $0
  require 'erb/main'

  ERbModule.run(ERbLight)
end


erb/compile.rb
# Tiny Tiny eRuby --- compiler
# 	Copyright (c) 1999-2000 Masatoshi SEKI 
#       You can redistribute it and/or modify it under the same term as Ruby.
#	$Id: compile.rb,v 1.1.1.1 2001/06/08 08:02:25 spc Exp $
# 	$Author: spc $

class ERbCompiler
  ERbTag = "<%% %%> <%= <%# <% %>".split
  private
  def self.is_erb_tag?(s)
    ERbTag.member?(s)
  end

  SplitRegexp = /(<%%)|(%%>)|(<%=)|(<%#)|(<%)|(%>)|(\n)/

  public
  def pre_compile(s, trim_mode)
    re = SplitRegexp
    if trim_mode.nil? || trim_mode == 0
      list = s.split(re)
    else
      list = []
      s.each do |line|
	line = line.chomp.split(re)
	line.shift if line[0]==''
	list += line
	unless ((trim_mode == 1 && line[-1] == '%>') ||
		(trim_mode == 2 && (is_erb_tag?(line[0])) && 
		 line[-1] == '%>'))
	  list.push("\n") 
	end
      end
    end
    list
  end

  def compile(s)
    list = pre_compile(s, @trim_mode)
    cmd = []
    cmd.concat(@pre_cmd)

    stag = nil
    content = []
    while (token = list.shift) 
      if token == '<%%'
	token = '<'
	list.unshift '%'
      elsif token == '%%>'
	token = '%'
	list.unshift '>'
      end
      if stag.nil?
	if ['<%', '<%=', '<%#'].include?(token)
	  stag = token
	  content.join.each do |line|
	    cmd.push("#{@put_cmd} #{line.dump}")
	  end
	  content = []
	else
	  content.push(token)
	end
      else
	if token == '%>'
	  content = content.join
	  case stag
	  when '<%'
	    cmd.push(content)
	  when '<%='
	    cmd.push("#{@put_cmd}((#{content}).to_s)")
	  when '<%#'
	    #	  cmd.push('=begin')
	    #	  cmd.push(content)
	    #	  cmd.push('=end')
	    cmd.push("# #{content.dump}")
	  end
	  stag = nil
	  content = []
	else
	  content.push(token)
	end
      end
    end
    content.join.each do |line|
      cmd.push("#{@put_cmd} #{line.dump}")
    end
    cmd.concat(@post_cmd)
    cmd.join("\n")
  end

  def initialize
    @trim_mode = nil
    @put_cmd = 'print'
    @pre_cmd = []
    @post_cmd = []
  end
  attr :trim_mode, true
  attr :put_cmd, true
  attr :pre_cmd, true
  attr :post_cmd, true
end



rubbs.rb
#
# RuBBS
#
# Copyright (C) 2001, All right reserved by TADA Tadashi <sho@spc.gr.jp>
# You can redistribute it and/or modify it under GPL2.
# $Revision: 1.22 $
#
RUBBS_VERSION = '1.0.7'

begin
	require 'gdbm'
	DBClass = GDBM
	DB_SUFFIX = '.gdb'
rescue LoadError
	begin
		require 'dbm'
		DBClass = DBM
		DB_SUFFIX = ''
	rescue LoadError
		require 'sdbm'
		DBClass = SDBM
		DB_SUFFIX = '.sdb'
	end
end
require 'cgi'
require 'nkf'
require 'erb/erbl'
require 'lock'

#
# enhanced String class
#
class String
	SALT_CHARSET = "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
	def String::salt
		result = '$1$'
		10.times do
			result << SALT_CHARSET[rand( 64 )]
		end
		result
	end

	def make_link
		return self.gsub( / /, "\001" ).
			gsub( /</, "\002" ).
			gsub( />/, "\003" ).
			gsub( %r|http[s]{0,1}://[\(\)%#!/0-9a-zA-Z_$@.&+-,'"*=;?:~-]+|, '<a href="\&">\&</a>' ).
			gsub( %r|ftp://[\(\)%#!/0-9a-zA-Z$-_@.&+-,'"*=;?:~]+|, '<a href="\&">\&</a>' ).
			gsub( %r|[0-9a-zA-Z_.-]+@[\(\)%!0-9a-zA-Z_$@.&+-,'"*-]+|, '<a href="mailto:\&">\&</a>' ).
			gsub( /&/, '&amp;' ).
			gsub( /\003/, '&gt;' ).
			gsub( /\002/, '&lt;' ).
			gsub( /\001/, '&nbsp;' ).
			gsub( /\t/, '&nbsp;' * 8 )
	end

	def quote
		quoted = ''
		self.each_line do |line|
			quoted += line.sub( /^\s*(>|&gt;).*\n*/, %|<span class="quote">#{line.chomp}</span>\n| )
		end
		return quoted
	end

	def nkf( charset )
		case charset
		when /iso-2022-jp/i
			self.to_jis
		when /euc-jp/i
			self.to_euc
		when /sjis/i, /shift_jis/i
			self.to_sjis
		else
			self.clone
		end
	end

	def native
		case $KCODE[0]
		when ?E, ?e
			self.to_euc
		when ?S, ?s
			self.to_sjis
		else
			self.clone
		end
	end
	def to_jis; NKF::nkf( '-m0 -jXd', self ); end
	def to_euc; NKF::nkf( '-m0 -eXd', self ); end
	def to_sjis; NKF::nkf( '-m0 -sXd', self ); end
end

#
# enhanced CGI class
#
class CGI
	def valid?( param, idx = 0 )
		self.params[param] and self.params[param][idx] and self.params[param][idx].length > 0
	end
end

#
# enhanced Time class
#
class Time
	def rfc1123_date
		self.clone.gmtime.strftime( '%a, %d %b %Y %X GMT' )
	end
end

#
# an Article
#
class Article
	attr_reader :name, :mail, :subject, :date, :remote, :contents, :reply_to

	def Article::decode( str )
		article = Article::new
		article.decode( str )
	end

	def initialize( name='', mail='', subject='(No Subject)', contents = '', nowrap = false, reply_to = '' )
		@name, @mail, @subject, @contents, @nowrap, @reply_to = name, mail, subject, contents, nowrap, reply_to
		@contents.sub!( /\n+\z/, '' )
		@date = Time.now.strftime( '%Y/%m/%d %H:%M:%S %Z' ).native
		@remote = ENV['REMOTE_ADDR']
		@status = "alive"
	end
	
	def equals?( article )
		if @name == article.name and
				@subject == article.subject and
				@contents.strip == article.contents.strip then
			true
		else
			false
		end
	end

	def decode( str )
		header = true
		str.each_line do |line|
			if header then
				case line
				when /^name: (.*)$/
					@name = $1
				when /^mail: (.*)$/
					@mail = $1
				when /^subject: (.*)$/
					@subject = $1
				when /^date: (.*)$/
					@date = $1
				when /^nowrap: true$/
					@nowrap = true
				when /^status: (.*)$/
					@status = $1
				when /^remote: (.*)$/
					@remote = $1
				when /^reply_to: (.*)$/
					@reply_to = $1
				else
					header = false if /^$/ =~ line
				end
			else
				@contents += line
			end
		end
		@contents.sub!( /\n+\z/, '' )
		self
	end

	def alive
		@status = 'alive'
	end

	def remove
		@status = 'remove'
	end

	def alive?
		@status == 'alive'
	end

	def encode
		return	"name: #{@name}\n" +
					"mail: #{@mail}\n" +
					"subject: #{@subject}\n" +
					"remote: #{@remote}\n" +
					"date: #{@date}\n" +
					"reply_to: #{@reply_to}\n" +
					"nowrap: #{@nowrap}\n" +
					"status: #{@status}\n" +
					"\n#{@contents}\n"
	end

	def html_contents
		c = @contents.make_link.quote
		if @nowrap then
			'<pre>' + c + '</pre>'
		else
			c.gsub( /\n/, "<br />\n" )
		end
	end
end

#
# a Board
#
class Board
	class BoardError < StandardError; end
	class BoardConfigError < BoardError; end
	class BoardBadOperation < BoardError; end

	attr_reader :name, :action, :encoding, :latest
	VIEW = 0
	MAINTENANCE = 1
	CONFIGURE = 2
	REMOVE = 3
	RESTORE = 4
	APPEND = 5
	REPLY = 6

	def initialize( cgi )
		begin
			@cgi = cgi
			@cgi_url = @cgi.script_name
			@name = File::basename( @cgi_url ).sub( /\..*$/, '' )
			
			# loading configuration
			load_conf

			# set action
			begin
				eval( "@action = #{@cgi.params['action'][0].upcase}" )
			rescue NameError
				@action = VIEW
			end
		rescue
			raise BoardConfigError::new( $!.to_s )
		end

		# other variables
		@top = 0
		@cookie = nil
		@m_msg = nil
		@m_serial = nil
	end

	def transaction
		raise BoardError::new( 'call with block' ) if not iterator?
		begin
			@db = DBClass::open( "#{@name}#{DB_SUFFIX}" )
			@latest = (@db['latest'] or '1').to_i
			@last_modified = @db['last_modified'] or Time::now.rfc1123_date
			yield
		ensure
			@db.close if @db
		end
	end

	def header( opt = nil )
		hash = {
			'type' => 'text/html',
			'charset' => @encoding,
		}
		opt.each do |k,v| hash[k] = v end if opt
		hash['Cache-Control'] = 'no-cache' if @action != VIEW
		hash['cookie'] = @cookie if @cookie
		@cgi.header( hash )
	end

	def maintenance
		if not @password or
				(@cgi.valid?( 'password' ) and @cgi.params['password'][0].crypt( @password ) == @password) then
			ERbLight::new( open_rhtml( '.conf.rhtml' ) ).result( binding ).nkf( @encoding )
		else
			raise BoardBadOperation::new( 'password unmatched' )
		end
	end

	def configure
		check_password do
			if @cgi.valid?( 'password' ) then
				if @cgi.params['password'][0] != @cgi.params['password2'][0] then
					@m_msg = @password_diff
					return maintenance
				else
					@new_password = @cgi.params['password'][0].crypt( String::salt )
				end
			else
				@new_password = @password
			end
			conf = ERbLight::new( open_rhtml( '.conf.rtxt' ) ).result( binding )
			File::open( "#{@name}.conf.new", 'w' ) do |f|
				f.write( conf )
			end
			File::delete( "#{@name}.conf~" ) if File::exist?( "#{@name}.conf~" )
			File::rename( "#{@name}.conf", "#{@name}.conf~" );
			File::rename( "#{@name}.conf.new", "#{@name}.conf" );
			load_conf
		end
		view
	end

	def remove
		check_password do
			if @cgi.valid?( 'serial' ) then
				serial = @cgi.params['serial'][0].sub( /^0+/, '' )
				if @db[serial] then
					article = Article::decode( @db[serial].native )
					if article.alive? then
						article.remove
						@db[serial] = article.encode.native
						@m_msg = @remove_success
					else
						@m_msg = @remove_fail
					end
				else
					@m_msg = @remove_fail
				end
				@m_serial = serial
			end
			return ERbLight::new( open_rhtml( '.conf.rhtml' ) ).result( binding ).nkf( @encoding )
		end
		raise BoardBadOperation::new( 'password unmatched' )
	end

	def restore
		check_password do
			if @cgi.valid?( 'serial' ) then
				serial = @cgi.params['serial'][0].sub( /^0+/, '' )
				if @db[serial] then
					article = Article::decode( @db[serial].native )
					if not article.alive? then
						article.alive
						@db[serial] = article.encode.native
						@m_msg = @restore_success
					else
						@m_msg = @restore_fail
					end
				else
					@m_msg = @restore_fail
				end
				@m_serial = serial
			end
			return ERbLight::new( open_rhtml( '.conf.rhtml' ) ).result( binding ).nkf( @encoding )
		end
		raise BoardBadOperation::new( 'password unmatched' )
	end

	def view
#		if @cgi.valid?( 'serial' ) then
#			serial = @cgi.params['serial'][0].to_i
#			ERbLight::new( open_rhtml( '.one.rhtml' ) ).result( binding ).nkf( @encoding )
#		else
			ERbLight::new( open_rhtml( '.view.rhtml' ) ).result( binding ).nkf( @encoding )
#		end
	end

	def form
		subject, content = '', ''
		serial, = @cgi.params['serial']
		if @cgi.valid?( 'serial' ) and (str = @db[serial] ) then
			a = Article::decode( str.native )
			subject = 'Re: ' + a.subject.sub( /^(re:\s*)+/i, '' )
			content = CGI::escapeHTML( a.contents.to_a.collect{|s|'>'+s}.join + "\n" )
		end
		name = @cgi.cookies['rubbs'][0] or ''
		mail = @cgi.cookies['rubbs'][1] or ''
		ERbLight::new( open_rhtml( '.form.rhtml' ) ).result( binding ).nkf( @encoding )
	end

	def post
		return if not @cgi.valid?( 'name' )
		name = @cgi.params['name'][0].native
		mail = @cgi.params['mail'][0].native
		subject = @cgi.valid?( 'subject' ) ? @cgi.params['subject'][0] : '(No Subject)'
		@reply_to = @cgi.valid?( 'reply_to' ) ? @cgi.params['reply_to'][0] : ''
		contents = @cgi.params['contents'][0].native
		nowrap = @cgi.valid?( 'nowrap' ) ? true : false
		article = Article::new( name, mail, subject.native, contents, nowrap, @reply_to )
		if @latest > 1 then # duplication check
			prev = Article::decode( @db[(@latest-1).to_s].native )
			return if prev and article.equals?( prev )
		end
		@db[@latest.to_s] = article.encode.native
		sendmail( @latest, article ) if not @mail_to.empty?
		@db['latest'] = (@latest += 1).to_s
		@db['last_modified'] = @last_modified = Time::now.rfc1123_date

		@cookie = CGI::Cookie::new( {
				'name' => 'rubbs',
				'value' => [name,mail],
				'path' => @cgi_url,
				'expires' => Time::now.gmtime + 24*60*60*30
		} )
	end

private

	def each_article
		raise BoardError::new( 'call with block' ) if not iterator?

		if @cgi.valid?( 'top' ) and not @cgi.valid?( 'ignore_top' ) then
			@top = @cgi.params['top'][0].to_i
			@top = @latest - 1 if @top >= @latest
		else
			@top = @latest - 1
		end

		@bottom = @top - @limit_per_page.to_i + 1
		@bottom = 1 if @bottom < 1

		@top.downto( @bottom ) do |serial|
			str = @db[serial.to_s]
			yield( serial, Article::decode( str.native ) ) if str
		end
	end

	def load_conf
		[	"conf/default.conf", "#{@name}.conf" ].each do |file|
			eval( File::readlines( file ).join )
		end
		[@keywords, @title, @header1, @header, @footer].each do |s|
			s.chomp! if s
		end
		lang = @encoding=='ISO-8859-1' ? 'en' : 'ja'
		eval( File::readlines( "conf/default.#{lang}.conf" ).join )
	end

	def open_rhtml( suffix )
		begin
			File::readlines( "#{@name}#{suffix}" ).join
		rescue Errno::ENOENT
			File::readlines( "conf/default#{suffix}" ).join
		end
	end

	def check_password
		if not @password or
				(@cgi.valid?( 'crypt' ) and @cgi.params['crypt'][0] == @password) then
			yield
		end
	end

	def sendmail( serial, article )
		return if not ($rubbs_smtp_host and $rubbs_smtp_from)

		case @lang
		when /^ja/
			name = to_mime( article.name )
			subject = to_mime( article.subject )
		else
			name = article.name
			subject = article.subject
		end
		mail = article.mail.length == 0 ? 'anonymous@unknown' : article.mail
		reply = @mail_reply
		if /^ja/ =~ @lang then
			encoding = 'ISO-2022-JP'
		else
			encoding = @encoding
		end
		contents = article.contents

		now = Time::now
		g = now.dup.gmtime
		l = Time::local( g.year, g.month, g.day, g.hour, g.min, g.sec )
		tz = (g.to_i - l.to_i) / 36
		date = now.strftime( "%a, %d %b %Y %X " ) + sprintf( "%+05d", tz )


		begin
			require 'net/smtp'
			message_id = "<#{@name}.#{serial}@#{Socket::gethostname.sub(/^.+?\./,'')}>"
			if not @reply_to.empty?
				@in_reply_to = "<#{@name}.#{@reply_to}@#{Socket::gethostname.sub(/^.+?\./,'')}>"
			end
			text = ERbLight::new( File::readlines( 'conf/mail.rtxt' ).join ).result( binding )
			text = text.to_jis if /^ja/ =~ @lang
			ENV['HOSTNAME'] = Socket::gethostname if not ENV['HOSTNAME'] # for Ruby 1.4's net/smtp
			Net::SMTP.start( $rubbs_smtp_host, 25 ) do |smtp|
				smtp.ready( $rubbs_smtp_from, @mail_to ) do |adapter| adapter.write( text ) end
			end
			File::open( 'mail.log', 'a' ) do |f|
				f.puts Time::now
				f.puts "succeed."
			end if $DEBUG
		rescue
			File::open( 'mail.log', 'a' ) do |f|
				f.puts Time::now
				f.puts $!
				f.puts $@.join( "\n" )
			end if $DEBUG
		end
	end

	def to_mime( str )
		"=?ISO-2022-JP?B?" + [str.to_jis].pack( 'm' ).gsub( /\n/, '' ) + "?="
	end
end

#
# RuBBS main class
#
class RuBBS
	def initialize
		begin
			eval( File::readlines( "rubbs.conf" ).join )
			$rubbs_smtp_host = @smtp_host if @smtp_host
			$rubbs_smtp_from = @smtp_from if @smtp_from
			
			cgi = CGI::new
			@bbs = Board::new( cgi )
			Lock::lock( "lock/#{@bbs.name}" ) do
				head = @bbs.header
				body = nil
				@bbs.transaction do
					case @bbs.action
					when Board::MAINTENANCE
						body = @bbs.maintenance
					when Board::CONFIGURE
						@bbs.configure
						update_html( @bbs )
						jump_html( @bbs )
					when Board::REMOVE
						body = @bbs.remove
						update_html( @bbs )
					when Board::RESTORE
						body = @bbs.restore
						update_html( @bbs )
					when Board::REPLY
						body = @bbs.form
					when Board::APPEND
						@bbs.post
						update_html( @bbs )
						jump_html( @bbs )
					else
						if cgi.valid?( 'serial' )
							body = @bbs.view
						elsif cgi.valid?( 'top' ) and not cgi.valid?( 'ignore_top' ) then
							body = @bbs.view
						elsif cgi.valid?( 'effect' ) and cgi.params['effect'][0] == 'yes' # making static HTML
							update_html( @bbs )
							jump_html( @bbs )
						else
							jump_html( @bbs )
						end
					end # case
				end # transaction
				if body then
					print head
					print body
				end
			end # lock
		rescue Lock::LockError 
			throw_html( cgi, 'conf/rubbs.lock.rhtml' )
		rescue Board::BoardBadOperation
			jump_html( @bbs )
		rescue # other errors
			throw_html( cgi, 'conf/rubbs.unknown.rhtml' )
		end
	end

	def update_html( bbs )
		File::open( "#{@path}#{bbs.name}.html", 'w' ) do |f|
			f.write( bbs.view )
		end
	end

	def jump_html( bbs )
		url = "#{@url}#{bbs.name}.html"
		url += "?post=#{bbs.latest-1}" if bbs.action == Board::APPEND
		print @bbs.header( { 'Location' => url} )
	end

	def throw_html( cgi, file )
		encoding = 'ISO-8859-1'
		lang = 'en_JS'
		if ENV['HTTP_ACCEPT_LANGUAGE'] then
			case ENV['HTTP_ACCEPT_LANGUAGE'].gsub( /\s+/, '' )
			when /^ja/
				encoding = 'ISO-2022-JP'
				lang = 'ja-JP'
			end
		end
		print cgi.header( { 'type' => 'text/html', 'charset' => encoding } )
		rhtml = File::readlines( file ).join
		print ERbLight::new( rhtml ).result( binding ).nkf( encoding )
	end
end

